src/hardware.rs

```diff
@@ -0,0 +1,261 @@
+
+const MEMORY_MAX: usize = 65536;
+const REGS_COUNT: usize = 10;
+
+/// Abstraction of the memory.
+/// It has 65,536 memory locations.
+pub struct Memory {
+    inner: [u16; MEMORY_MAX],
+}
+
+impl Memory {
+    pub fn new() -> Self {
+        Self {
+            inner: [0; MEMORY_MAX],
+        }
+    }
+
+    /// Sets a new val in the specified memory address
+    ///
+    /// ### Arguments
+    ///
+    /// - `mem_address`: A generic that can be converted into an usize and represents the memory address to write on.
+    /// - `new_val`: The new val to write on the specified memory address.
+    ///
+    /// ### Returns
+    ///
+    /// A Result indicating whether the operation failed or not. This operation can fail if the address
+    /// to write on is an invalid one. An address is invalid if it is not in [0, 65535].
+    pub fn write<T: Into<usize>>(&mut self, mem_address: T, new_val: u16) -> Result<(), VMError> {
+        let index: usize = mem_address.into();
+        if let Some(val) = self.inner.get_mut(index) {
+            *val = new_val;
+            return Ok(());
+        }
+        Err(VMError::InvalidIndex(String::from(
+            "Invalid index when writing on memory",
+        )))
+    }
+
+    /// Reads a memory address. If the memory address to read is the one that stores
+    /// the KeyboardStatus, then it updates the KeyboardData address in the memory
+    /// by writing the character that was read from standard input.
+    ///
+    /// ### Arguments
+    ///
+    /// - `addr`: An u16 representing the memory address to read from.
+    ///
+    /// ### Returns
+    ///
+    /// A Result containing the data in the memory address, or a VMError if
+    /// the operation failed. The operation can fail if writing in the memory fails
+    /// (writtings are done when a character was read from stdin) or because
+    /// the address is an invalid one and is not in the range [0, 65535].
+    pub fn read(&mut self, addr: u16) -> Result<u16, VMError> {
+        if addr == MemoryRegister::KeyboardStatus {
+            self.write(MemoryRegister::KeyboardStatus, 1 << 15)?;
+            let mut reader = stdin();
+            let buffer = getchar(&mut reader)?;
+            let char: u16 = buffer[0].into();
+            self.write(MemoryRegister::KeyboardData, char)?;
+        }
+        // Get the value
+        let index: usize = addr.into();
+        if let Some(val) = self.inner.get(index) {
+            return Ok(*val);
+        }
+        Err(VMError::InvalidIndex(String::from(
+            "Invalid index when writing on memory",
+        )))
+    }
+}
+
+/// Abstraction of a single register.
+/// We have:
+/// - 8 general purpose registers (R0-R7)
+/// - 1 program counter register (PC)
+/// - 1 condition flags register (COND)
+#[derive(Clone, Copy)]
+pub enum Register {
+    R0,
+    R1,
+    R2,
+    R3,
+    R4,
+    R5,
+    R6,
+    R7,
+    PC,
+    Cond,
+}
+
+impl Register {
+    fn index(&self) -> usize {
```

Comment:
to avoid adding this `index` function you can add `#[repr(u8)]` to the type and then just use, for example:

```rust
let reg = Register::R2;
println!("the register number is {}", reg as u8);
```